### ISSUE TITLE
How to remove zero-init GPU kernel?

### ISSUE THREAD

**Issue by mcourteaux** [halide/Halide#6449]
Let's say I have a function defined purely to be zero-initialized and then add an update stage using `func += ...` (with an RDom). This now produces two kernels: the first just setting the entire buffer to zero. Then a second kernel doing the actual calculation. I want to "inline" the zero-assignment into the second kernel to get rid of the first one. How can I do that?

```cpp
Func f;
f(k) = 0.0f;
f(k) += some_other_func(k, rdom);
```

What schedule do I need to produce one kernel that just computes this sum?

**Comment by mcourteaux** [comment]
Behavior I'm seeing is that the second kernel loads the zero-inited memory into a register. Then accumulates it nicely in the register, and then finally writes the result once. So that's all nice except for the load from memory, which is certain to yield zero.

Loading zero happens into `%f6` from address in `%rd3` in the following screenshot. You can nicely see the accumulation happening in the `fma` instruction (it's a more complicated kernel of course than my simplified question). Then in the end, one final write back.
![image](https://user-images.githubusercontent.com/845012/143864415-3cd280ff-e512-4e00-be6d-610be6d27996.png)

**Comment by steven-johnson** [comment]
This strikes me as (yet another) example of something it would be nice to have a set of "Halide Tips & Tricks" documented somewhere -- in a GitHub wiki, or (probably better) in a version-controlled Markdown doc in the repo, etc -- if anyone wants to take it upon themselves to start such a thing, I'd be happy to review it :-)

**Comment by abadams** [comment]
It could be added to tutorial lesson 19, but perhaps we need something more like a page of fast answers to simple questions.

**Comment by mcourteaux** [comment]
I would love to contribute that. I'm thinking so much about making some writeups or video about what I learned. I love Halide, but I'm spending A LOT of time figuring out how to do proper scheduling. The tutorials are very limited currently. Documentation for all the scheduling directives is sometimes too vague or abstract. I think the people who wrote the schedules for the /apps/ could be documenting them better to learn about their reasoning when making these schedules.  Now I find patterns with double `.rfactor()` directives or stuff with double `.in()` directives. I have no clue what `.in()` is practically used for, let alone being used twice. There is so much I still have to learn, despite working with this library for a year now.

**Comment by abadams** [comment]
Videos or write-ups would be much appreciated. Halide's biggest weakness is that scheduling is really hard.

**Comment by abadams** [comment]
Also the app schedules are mostly me. I'll try to document my thinking better the next time I touch one.

**Comment by steven-johnson** [comment]
> I'm spending A LOT of time figuring out how to do proper scheduling

If it's any consolation, I've been helping maintain Halide for years now and I *still* have a hard time with nontrivial schedules.

**Comment by mcourteaux** [comment]
I can't get option 3 to work, which is what seems ideal for my scenare, where I want to provide both schedules for CUDA and CPU-AVX2. If I use the anonymous function approach using `sum()`, it produces the CUDA kernel I want (I think), but then I can't provide a schedule for CPU anymore that does the update. I want to vectorize over `k`, which seems to be impossible as I cannot access the anonymous function to specify its schedule.

So maybe my minimal example was too simple and allowed a hack that isn't applicable in my case. Let's try again:

```cpp
Func f;
f(k) = 0.0f;
f(k) += some_other_func(k, rdom);

Func g;
g(k) = f(k) * f(k);
```

By the way, if it's fine with you, I can throw everything I'm seriously struggling with in an issue here on GitHub, and we tag it with FAQ or something. My project is a serious application for my PhD, so I'm sticking with it for a while, and I'm trying to outperform TensorFlow by a lot.

**Comment by abadams** [comment]
The pattern (assuming you want f and g as separate kernels) is:

```
Func f;
f(k) = 0.0f;
f(k) += some_other_func(k, rdom);

Func g;
g(k) = f(k) * f(k);

f.compute_at(f.in(), k);
f.in().compute_root()...

g.compute_root()...
```

**Comment by mcourteaux** [comment]
Okay, let's try this :)

**Comment by mcourteaux** [comment]
I'm stuck again. It at least compiles, but the schedule goes bananas. I'll paste actual code now, because I don't want to simplify things, and then miss details.

First of all, the definitions:

```cpp
        RDom domCltmm; // Reduction domain over lower triangular matrix.
        domCltmm = Halide::RDom{0, adims, "domCltmm"};
        domCltmm.where(domCltmm <= r);

        normDiff(s, k, r) = 0.0f;
        normDiff(s, k, r) += invCholR(k, r, domCltmm) * diff(s, k, domCltmm);

```

Starting without any of the tricks here, using two kernels (the pure and the update), I get this loopnest:
```cpp
    produce normDiff:
      gpu_block sample.so<Default_GPU>:
        gpu_block kernel<Default_GPU>:
          gpu_block row in [0, 6]<Default_GPU>:
            gpu_thread sample.si in [0, 63]<Default_GPU>:
              normDiff(...) = ...

      gpu_block row in [0, 6]<Default_GPU>:
        gpu_block kernel<Default_GPU>:
          gpu_block sample.so<Default_GPU>:
            gpu_thread sample.si in [0, 63]<Default_GPU>:
              for domCltmm in [0, 6]:
                normDiff(...) = ...
```
Pretty clean, but a waste of the first kernel. The schedule for above loopnest was this:
```cpp
        normDiff
            .compute_root()
            .split(s, so, si, 64)
            .reorder(si, r, so)
            .gpu_blocks(so, k, r)
            .gpu_threads(si)
            .update()
            .split(s, so, si, 64)
            .reorder(si, so, k, r)
            .gpu_blocks(so, k, r)
            .gpu_threads(si)
```
---
Now, trying your suggestion, I do this:

```cpp
        normDiff.compute_at(normDiff.in(), si);
        normDiff.in().compute_root()
            .split(s, so, si, 64)
            .reorder(si, so, k, r)
            .gpu_blocks(so, k, r)
            .gpu_threads(si)
            ;
```
Produces this loop nest:
```cpp
    produce normDiff_global_wrapper:
      gpu_block row in [0, 6]<Default_GPU>:
        gpu_block kernel<Default_GPU>:
          gpu_block sample.so<Default_GPU>:
            gpu_thread sample.si in [0, 63]<Default_GPU>:
              produce normDiff:
                for row in [0, 6]:
                  for kernel:
                    for sample:
                      normDiff(...) = ...
                for row in [0, 6]:
                  for kernel:
                    for sample:
                      for domCltmm in [0, 6]:
                        normDiff(...) = ...
              consume normDiff:
                normDiff_global_wrapper(...) = ...
```

So you can see that, although I do `normDiff.compute_at(normDiff.in(), si)` it for some reason repeats the three nested for loops in the `produce normDiff`. I want a loopnest like this (just edited here in the text-editor, this is not output):
```cpp
      gpu_block row in [0, 6]<Default_GPU>:
        gpu_block kernel<Default_GPU>:
          gpu_block sample.so<Default_GPU>:
            gpu_thread sample.si in [0, 63]<Default_GPU>:
              normDiff(...) = 0.0f
              for domCltmm in [0, 6]:
                normDiff(...) = ...
```

**Comment by abadams** [comment]
Check the final .stmt. It's quite possible those three unwanted nested loops have extent 1. print_loop_nest can unfortunately be misleading because it happens before bounds inference.

**Comment by mcourteaux** [comment]
I disabled the GPU scheduling directives to get a look at the statement file. Quite disastrous indeed:

![image](https://user-images.githubusercontent.com/845012/144095028-1dd68afa-62c1-448b-9856-f2b28c4cabac.png)

The PTX code reflects this madness if I turn on the gpu-directives. I have witnessed similar behavior many times before, and start thinking that it's a bug, if it isn't a very counter-intuitive side effect of how Halide scheduling works.

**Comment by abadams** [comment]
It looks like you may have a .bound() directive on normDiff. That forces the loop nest to be the given size. Try just applying it to the wrapper instead.

**Comment by mcourteaux** [comment]
Aha, so `.bound()` in combination with `.compute_at()` causes a forced full bounds evaluation of that function, effectively blowing up the complexity of the result (big-O wise complexity)? Well spotted! Thanks for your amazing help. I had these bounds set to every function as that made a lot of the address calculations simpler. Usually it subtracts the `.min.0` from everything, and it becomes complex and unnecessary, because I know I'll use min = 0. Setting the bounds fixed this in quite a few cases, so I just made a habit of putting that everywhere.

Nice! Applying these to the wrapper makes sense indeed. I'm wondering tho: Now I put the bounds in the `generate()`, because I thought of them as being independent of the schedule. However, introducing in the schedule a wrapper with `.in()` makes that assumption invalid. What do you think here? Should the creation of a wrapper (by using `.in()`) automatically move the bounds information to the wrapper instead?

**Comment by abadams** [comment]
There are probably applications of .bound that are distinct to the wrapper and the wrappee, so I wouldn't want to move bound information on vars over. Though in the other issue I just opened (#6459) I do suggest moving buffer constraint information over to the wrapper.

.bound should be thought of as part of the schedule, IMO. It depends on the compute_at.

**Comment by mcourteaux** [comment]
Yes!! It works with the bounds on the wrapper. Thanks for your help. What do you think about the FAQs label?

Here init to 0.0f, and then the loop to accumulate results. Beautiful!
![image](https://user-images.githubusercontent.com/845012/144262951-9d6672f6-93f7-408e-98d1-01dd6678dfe5.png)

**Comment by abadams** [comment]
Done.

**Comment by mcourteaux** [comment]
Cool. As this is not a bug in Halide, I think we can close this issue. Although I think a warning would be nice if your schedule explodes the computation complexity. Like in my case, I went from O(num_kernels * num_coords * 7 * 7) to O(num_kernels * num_coords * 7 * 7 * num_kernels * num_coords * 7 * 7). I don't know anything about the scheduling internals, so I don't know if this makes sense and is feasible.

**Comment by abadams** [comment]
Closed in favor of #6460
